### ISSUE TITLE
Demonstrate perfect restart for full phyics applications.

### ISSUE THREAD

**Issue by spdomin** [NaluCFD/Nalu#29]
A restarted simulation does not provide **exactly** the same results. For example, if I run a simulation from step 1:100 and then restart at step 50, the steps are really (really) close, however, not perfect. This suggests a subtle start-up issue with restart that should be resolved.

@NaluCFD/core, I need someone who will lead this task. Might be good for a new developer..

The recent dgNonConformalThreeBlade regression test is the target test problem. However, that test is probably too complex to start with since it has mesh motion, second order time integration (with adaptive time step), etc.

I would start with a simple heat conduction case and move up in complexity.

**Comment by spdomin** [comment]
I just noticed that in my moving mesh simulations, the population of the velocity relative to mesh is not computed until the realm::integrated_realm(). Making this change caused some diffs in restart (which I now thing are more consistent).

The change was simple and something that I will push very soon. I still think that all is not perfect though..

```
diff --git a/src/LowMachEquationSystem.C b/src/LowMachEquationSystem.C
index dfb6cc1..74210ec 100644
--- a/src/LowMachEquationSystem.C
+++ b/src/LowMachEquationSystem.C
@@ -855,6 +855,7 @@ MomentumEquationSystem::initial_work()
 
   // proceed with a bunch of initial work; wrap in timer
   const double timeA = NaluEnv::self().nalu_time();
+  realm_.compute_vrtm();
   compute_projected_nodal_gradient();
   compute_wall_function_params();
   tviscAlgDriver_->execute();

```

**Comment by spdomin** [comment]
I need this to be reassigned....

**Comment by sayerhs** [comment]
Stefan, 

Mike asked me if I could look into this one. Can you confirm that this task is still open? 

I started with `femHC` and the restart worked correctly on it. Then I moved to `steadyTaylorVortex` and I am starting to see differences. Before I started digging into this case, I wanted to check to see if there was an intermediate (simpler?) case that might make sense to check before tackling `steadyTaylorVortex`.

**Comment by spdomin** [comment]
Awesome.. Please do take this on! The steadyTaylorVortex is probably the simplest case around; 2D and only uv_p - all inflow bcs. Moreover, fixed time step and first order time integration.

I would concentrate on that test case..

**Comment by sayerhs** [comment]
OK. Just saw this, but some updates with `steadyTaylorVortex`. I found that the differences are arising from the `ProjectedNodalGradientEquationSystem` fields not being in the restart file. By using `realm_.augment_restart_variable_list` on `dofName_` and `deltaName_`. The restart differences reduce to `4e-16` in the Mean System Norm. 

Following this up with a modification in `LowMachEquationSystem.C:568` to 

```
if (!realm_.restarted_simulation()) continuityEqSys_->compute_projected_nodal_gradient();
```

The differences go down to `1e-19` in the mean system norm, sufficient to pass the tolerance test with `.norm.gold` file for the timesteps considered. 

Since `pressure` is already in the restart, I am yet unclear why I've to specify `dpdx` in the restart (for the `steadyTaylorVortex` case). May be you an explanation why.

**Comment by spdomin** [comment]
dpdx is, in fact, required within the first momentum solve. I thought, however, that as an initialization "step", we process dpdx...

  ```
if ( isInit_ ) {
    timeA = NaluEnv::self().nalu_time();
    continuityEqSys_->compute_projected_nodal_gradient();
    continuityEqSys_->computeMdotAlgDriver_->execute();
    timeB = NaluEnv::self().nalu_time();
    continuityEqSys_->timerMisc_ += (timeB-timeA);
    isInit_ = false;
  }
``` 
The above code means that by the time we process the first momentum solve, dpdx should be correct whether or not we provided it within the restart. 

So, there is a mystery here... At the very least, we should never need the delta in the restart...

**Comment by sayerhs** [comment]
@spdomin Question: Will the following code update copy over a field everywhere (i.e., on interior nodes as well as boundary nodes)? I am calling this method after the call for `computeMdotAlgDriver_`, but still see some discrepancy in the computed `mass_flow_rate` ( on the order of ~1e-11) when compared to the baseline. I see a `CopyFieldAlgorithm` class, but I am unsure that is applicable here. There is a `copy_dpdx_after_restart` that is used to reset `dpdx` just before the `computeModAlgDriver_` call within the `isInit` if statement. 

```
void
ContinuityEquationSystem::save_dpdx_for_restart()
{
  stk::mesh::MetaData & meta_data = realm_.meta_data();
  const int nDim = meta_data.spatial_dimension();

  // field that we need
  VectorFieldType *dpdx = dpdx_;
  VectorFieldType *dpdxRst = dpdxRestart_;

  // selector (everywhere dpdx lives) and node_buckets 
  stk::mesh::Selector s_nodes = stk::mesh::selectField(*dpdx);
  stk::mesh::BucketVector const& node_buckets =
    realm_.get_buckets( stk::topology::NODE_RANK, s_nodes );

  for ( stk::mesh::BucketVector::const_iterator ib = node_buckets.begin() ;
        ib != node_buckets.end() ; ++ib ) {
    stk::mesh::Bucket & b = **ib ;
    const stk::mesh::Bucket::size_type length   = b.size();
    double * dp = stk::mesh::field_data(*dpdx, b);
    double * dpRst = stk::mesh::field_data(*dpdxRst, b);

    for ( stk::mesh::Bucket::size_type k = 0 ; k < length ; ++k ) {
      const int offSet = k*nDim;
      for ( int j = 0; j < nDim; ++j ) {
        dpRst[offSet+j] = dp[offSet+j];
      }
    }
  }
}
```

**Comment by spdomin** [comment]
Yes, this is a good bucket loop that will copy dpdx to dpdx_restart at all nodes. However, you probably want to use the field function call,

`  field_copy(realm_.meta_data(), realm_.bulk_data(), dpdx_, dpdxRst_, realm_.get_activate_aura());`

As for other nuances, we may need the same procedure with the velocity field since reconstructing dot requires the velocity field that has not yet been projected. I think that this is the reason why we always (in other codes) save off dot to restart.

I have a question into the STK team, however, our own @alanw0 can probably help as well.

**Comment by alanw0** [comment]
I hadn't previously noticed that nalu had its own field_copy function. I was about to reply and recommend using stk's field_copy function. There is a subtle difference though: Nalu's function takes a 'auraIsActive' flag, which if false, causes the field to not be copied for ghosted entities. (Including custom-ghosted entities.) I'm not sure if this is intended or desired behavior, but it could result in diffs if the behavior is changed, depending on where the nearest communicate_field_data call is that syncs the field in question.

Another note: if the 'auraIsActive' flag is true, nalu's field_copy function 'and's the field selector with the universal part. I don't think that has any effect, as everything is in the universal part by definition. (I.e., anding that selector would not exclude any entities.)

**Comment by alanw0** [comment]
I just replied to the stk help ticket regarding edge fields in restart. I added Greg Sjaardema to the cc list for the ticket, and asked him whether edges can be written to an exodus file for a 3D mesh. If they can, then supporting field output for those edges will probably be pretty easy. If they can't, then it will be harder.

**Comment by spdomin** [comment]
I am not sure if this code belongs part of this ticket or now, however, for completeness, this is the selector logic in our FieldFunctions class: 

 ```
 // decide on selector
  const stk::mesh::Selector selector = auraIsActive 
    ? metaData.universal_part() &
    stk::mesh::selectField(xField) &
    stk::mesh::selectField(yField)
    : (metaData.locally_owned_part() | metaData.globally_shared_part()) &
    stk::mesh::selectField(xField) &
    stk::mesh::selectField(yField);
```
The intent of the FieldFunction copy is to not concern itself with ghosted entities through custom ghosting - at least that was my recollection on the design. However, at present I am not sure why I was concerned over the aura aspect of field functions? Also, below the universal_part is "&" only with a selector with fields. So, if a field is not defined on a part, e.g., block_1, then I think the selector would not include it.

In general, we always call the communicate_field_data() specially when dealing with parallel communication of custom-ghosted fields, e.g.,

```
  // parallel communicate ghosted entities
  if ( NULL != realm_.nonConformalManager_->nonConformalGhosting_ )
    stk::mesh::communicate_field_data(*(realm_.nonConformalManager_->nonConformalGhosting_), ghostFieldVec_); 

```
 ```
 // decide on selector
  const stk::mesh::Selector selector = auraIsActive 
    ? metaData.universal_part() &
    stk::mesh::selectField(xField) &
    stk::mesh::selectField(yField)
    : (metaData.locally_owned_part() | metaData.globally_shared_part()) &
    stk::mesh::selectField(xField) &
    stk::mesh::selectField(yField);
```

**Comment by spdomin** [comment]
I see the point about universal not being required here... 

Part of the complexity with the core Nalu algorithm is that it looks like a mix between Jacobi and gauss-seidel updates. As such, its always user beware. I think that these field update methods were really targeting the linear system updates. These are good comments though and I will take a second look at the core design..

**Comment by spdomin** [comment]
Conclusion: The main issue with 100% clean restarts was that we found that the fact that mass flow rate at interior points was not set as a restart field caused diffs. For element-based schemes, we can add this field as a restart field. However, for edge-based, we require added STK functionality.

We evaluated reconstructing the proper Mdot, however, that required adding specialty fields, e.g., velocity_before_projection and dpdx_before_projection which added complexity to the design. 

The end path forward is to support integration point data restart for both edge- and element-based schemes. We can probably flip the switch for element-based now if we like.

In my restarts, I found that if I am converging within the time step, then restarted simulation converge to the same state as the first simulation. For now, we can close this and transition to a new feature request at a later date.
